state/storage_test.go

```diff
@@ -687,6 +689,139 @@ func TestGlobalSnapshot(t *testing.T) {
 	}
 }
 
+func TestAllJoinedMembers(t *testing.T) {
+	assertNoError(t, cleanDB(t))
+	store := NewStorage(postgresConnectionString)
+	defer store.Teardown()
+
+	alice := "@alice:localhost"
+	bob := "@bob:localhost"
+	charlie := "@charlie:localhost"
+	doris := "@doris:localhost"
+	eve := "@eve:localhost"
+	frank := "@frank:localhost"
+
+	// Alice is always the creator and the inviter for simplicity's sake
+	testCases := []struct {
+		Name                  string
+		InitMemberships       [][2]string
+		AccumulateMemberships [][2]string
+		RoomID                string // tests set this dynamically
+		WantJoined            []string
+		WantInvited           []string
+	}{
+		{
+			Name:                  "basic joined users",
+			InitMemberships:       [][2]string{{alice, "join"}},
+			AccumulateMemberships: [][2]string{{bob, "join"}},
+			WantJoined:            []string{alice, bob},
+		},
+		{
+			Name:                  "basic invited users",
+			InitMemberships:       [][2]string{{alice, "join"}, {charlie, "invite"}},
+			AccumulateMemberships: [][2]string{{bob, "invite"}},
+			WantJoined:            []string{alice},
+			WantInvited:           []string{bob, charlie},
+		},
+		{
+			Name:                  "many join/leaves, use latest",
+			InitMemberships:       [][2]string{{alice, "join"}, {charlie, "join"}, {frank, "join"}},
+			AccumulateMemberships: [][2]string{{bob, "join"}, {charlie, "leave"}, {frank, "leave"}, {charlie, "join"}, {eve, "join"}},
+			WantJoined:            []string{alice, bob, charlie, eve},
```

Comment:
Is this correct? Shouldn't charlie have left?

Reply:
? Charlie leaves then charlie joins later on?